pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -137,6 +163,108 @@ func (grb *globalRoleBindingLifecycle) deleteAdminBinding(obj *v3.GlobalRoleBind
+		return fmt.Errorf("unable to list clusters when reconciling globalRoleBinding %s: %w", globalRoleBinding.Name, err)
+	}
+
+	var missedClusters bool
+	for _, cluster := range clusters {
+		// inheritedClusterRoles only apply on non-local clusters, so skip the local cluster
+		if cluster.Name == localClusterName {
+			continue
+		}
+		missingRTs, err := grb.purgeCorruptRoles(globalRole.InheritedClusterRoles, cluster, globalRoleBinding)
+		if err != nil {
+			logrus.Errorf("unable to purge roles for cluster %s and grb %s, some permissions may be missing: %s", cluster.Name, globalRoleBinding.Name, err.Error())
+			missedClusters = true
+			continue
+		}
+		// at this point, the only remaining items are roleTemplates that we don't have a CRTB for in this cluster
+		for _, wantRT := range missingRTs {
+			// create a crtb in the backing namespace for the cluster
+			logrus.Infof("creating backing crtb for grb %s in cluster %s for roleTemplate %s", globalRoleBinding.Name, cluster.Name, wantRT)
+			_, err = grb.crtbClient.Create(&v3.ClusterRoleTemplateBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "crtb-grb-",
+					Namespace:    cluster.Name,
+					// the owner ref needs to be mutable by the k8s garbage collector but we need
+					// a way to identify what CRTBs are from GRBs unambiguously for validation
+					Labels: map[string]string{
+						grbOwnerLabel: globalRoleBinding.Name,
+					},
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: v3.GlobalRoleBindingGroupVersionKind.GroupVersion().String(),
+							Kind:       v3.GlobalRoleBindingGroupVersionKind.Kind,
+							Name:       globalRoleBinding.Name,
+							UID:        globalRoleBinding.UID,
+						},
+					},
+				},
+				ClusterName:        cluster.Name,
+				RoleTemplateName:   wantRT,
+				UserName:           globalRoleBinding.UserName,
+				GroupPrincipalName: globalRoleBinding.GroupPrincipalName,
+			})
+			// we don't immediately return so that we can create as many CRTBs as we can
+			if err != nil {
+				logrus.Errorf("failed to create crtb for globalRoleBinding %s in cluster %s: %s", globalRoleBinding.Name, cluster.Name, err.Error())
+				missedClusters = true
+			}
+		}
+	}
+	if missedClusters {
+		return fmt.Errorf("unable to create backing crtbs for globalRoleBinding %s, some permissions may be missing", globalRoleBinding.Name)
+	}
+	return nil
+}
+
+// purgeCorruptRoles removes any CRTBs which were created for this role in the past, but are no longer valid. Returns
+// a slice of the roles that were in wantRTs but are not yet covered by the existing CRTBs
+func (grb *globalRoleBindingLifecycle) purgeCorruptRoles(wantRTs []string, cluster *v3.Cluster, binding *v3.GlobalRoleBinding) ([]string, error) {
+	currentRTs := map[string]struct{}{}
+	for _, wantRT := range wantRTs {
+		currentRTs[wantRT] = struct{}{}
+	}
+	currentCRTBs, err := grb.crtbCache.GetByIndex(crtbGrbOwnerIndex, fmt.Sprintf("%s/%s", cluster.Name, binding.Name))
```

Comment:
Please rename the method receiver from `grb` to `lifecycle` or something like that. You already receive a `grb` (under binding), it's confusing.